Make user bounding box updates more efficient

MichaelBuessemeyer opened the pull request:
This PR add more update actions for user bounding boxes. This makes it more efficient to update a single bounding box (including adding and deleting). Moreover, the version restore view can display more details on what was actually changed (e.g. color of the bounding box).

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- create/mutate bounding boxes
- save
- reload and check that bboxes look like before

### TODOs:
- [x] if code is kept: make saga code DRY in volumetracing_saga and skeletontracing_saga
- [x] use `updateUserBoundingBoxVisibility` actions for visibility updates 
    - [x] frontend
    - [x] backend
- [x] consider renaming actions (no "action" in name field AND new action name, e.g. `updateUserSpecificBoundingBox[Volume|Skeleton]`)
    - [x] frontend (done: no action in name field)
    - [x] backend

### Issues:
- fixes [#7604](https://github.com/scalableminds/webknossos/issues/7604)

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

fm3 commented:
@MichaelBuessemeyer How would you feel about three new actions, Add/Delete/Update bbox? That way we don’t need an update action for every property.

philippotto commented:
I'm also in favor of having less update action types. I think we should aim for update actions where not all properties have to be listed. For example, one could have `UpdateBoundingBox(name="new name")` and `UpdateBoundingBox(color="new color")`. then, we only have one class for the update, but the sets of changed keys can be different.

I think, we already had an issue for this, but I cannot find it anymore. from what I remember, the above is a bit tricky, because scala cannot distinguish between `undefined` and simply not listed in a JSON? or something like this? however, we should be able to work around this somehow (e.g., list the keys explicitly that should be set).

fm3 commented:
You are probably referring to https://github.com/scalableminds/webknossos/issues/6657 – there we didn’t consider this for update actions but it might just work there too.

knollengewaechs commented:
what timestamp should be put into the updateAction? the moment the first frontend action was fired or the moment when the annotation is saved?

knollengewaechs commented:
> however, we should be able to work around this somehow (e.g., list the keys explicitly that should be set).
my initial solution is to call the frontend updateAction with a `Partial<UserBoundindBox>` and when sending this to the backend, explicitly include the props in the partial as a list. if anyone thinks that this is a bad idea, please let me know!

knollengewaechs commented:
@MichaelBuessemeyer I updated the frontend code, so that new the new updateActions are called. I tested it a bit with the old backend code, and it seems like the right actions are dispatched if the bounding boxes change. Let me know if the code seems buggy in any way!

knollengewaechs commented on frontend/javascripts/oxalis/model/sagas/skeletontracing_saga.ts:
```diff
@@ -607,6 +613,59 @@ export const cachedDiffTrees = memoizeOne((tracingId: string, prevTrees: TreeMap
+      continue;
+    }
+    const diffBBox = Utils.diffObjects(currentBbox, prevBbox);
+    //TODO_C remove
```
remove

MichaelBuessemeyer commented:
> @MichaelBuessemeyer I updated the frontend code, so that new the new updateActions are called. I tested it a bit with the old backend code, and it seems like the right actions are dispatched if the bounding boxes change.

Nice :+1: awesome. Thanks a lot :pray:
I changed the frontend code a little to not return single elemented lists of user bounding boxes for the add actions. See c680cf285877a430c17ee20c8cd8eef452597b5b. This was necessary for further testing.

>  Let me know if the code seems buggy in any way!

Yeah sorry there is a little buggy / unexpected behaviour: 
When the bounds of a bbox are changed the changed bounds are not send as a bounding box object with `topLeft, width, height, depth` but as changes in the `min max` props. See the screenshot below (taken from the network tab showing a update actions transaction dispatched to the server):
![image](https://github.com/user-attachments/assets/64557513-7215-4a8c-887b-f81fbc23b9b8)
Could you please adjust the frontend to always send the whole bounds as an object (`topLeft, width, height, depth`) in case some of them were updated? Thanks a lot :pray: 

Moreover, I noticed that the frontend seems to send "empty" updates: The `updatedPropKeys` and `updatedProps` are empty. This is also shown in the screenshot in the first array entries value. Could you please investigate this and try to avoid such empty update? 

Moreover, the cherry on top would be that multiple consecutive bounding box updates to the same bounding box would be combined into a single action. Something like this is already done by the frontend before sending the update actions. See `compact_save_queue.ts` ll. 100ff. But I am unaware how easy this is to implement. So maybe that's a follow up :thinking:?

knollengewaechs commented:
@MichaelBuessemeyer thanks for the reports! I will work on this now and try to fix the three problems.

knollengewaechs commented:
let me know if you notice anything else. during my testing, adding, updating and deleting bboxes seems to work fine :)

MichaelBuessemeyer commented:
For reference: For now blocked. See https://scm.slack.com/archives/C5AKLAV0B/p1744295336264969

fm3 commented on webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/annotation/UpdateActions.scala:
```diff
@@ -54,16 +54,20 @@ object UpdateAction {
+          case "deleteUserBoundingBoxVolumeAction" => deserialize[DeleteUserBoundingBoxVolumeAction](jsonValue)
           case "updateUserBoundingBoxesInVolumeTracing" =>
             deserialize[UpdateUserBoundingBoxesVolumeAction](jsonValue)
-          case "updateUserBoundingBoxVisibilityInVolumeTracing" =>
```
Take care not to mix the scala class names vs the "name" field (string above). This is different for most actions. I think the "name" usually does not end on "Action"

knollengewaechs replied:
yes, I was talking about the "name" field. maybe we have to update the names of the actions that are introduced with this PR because they all end on "Action" :sweat_smile:

knollengewaechs commented on frontend/javascripts/oxalis/model/sagas/update_actions.ts:
```diff
@@ -38,11 +41,29 @@ export type CreateSegmentUpdateAction = ReturnType<typeof createSegmentVolumeAct
 export type DeleteSegmentDataUpdateAction = ReturnType<typeof deleteSegmentDataVolumeAction>;
-type UpdateUserBoundingBoxesInSkeletonTracingUpdateAction = ReturnType<
-  typeof updateUserBoundingBoxesInSkeletonTracing
+export type AddUserBoundingBoxSkeletonAction = ReturnType<
```
add In

fm3 reviewed (commented):
Thanks for tackling this! I think the backend looks pretty good :)

I’m not super happy with the updatedPropKeys as String list. Maybe it would be nicer to have an additional boolean per property in the json? Since you check for all keys by name in the if/elses anyway, I’d say the booleans should not mean much more code or complexity. If possibly, I’d like to avoid the strings. What do you think?

knollengewaechs commented:
I agree that this is probably not much overhead!

knollengewaechs commented on frontend/javascripts/viewer/model/sagas/skeletontracing_saga.ts:
```diff
@@ -607,6 +615,71 @@ export const cachedDiffTrees = memoizeOne((tracingId: string, prevTrees: TreeMap
+    const diffBBox = Utils.diffObjects(currentBbox, prevBbox);
+    if (_.isEmpty(diffBBox)) continue;
+    const changedKeys = Object.keys(diffBBox);
+    if (changedKeys.includes("isVisible")) {
+      yield updateBBoxVisibilityAction(currentBbox.id, currentBbox.isVisible, tracingId);
+      continue;
+    }
```
I disagree. the visibility is always updated on its own

philippotto commented on frontend/javascripts/viewer/model/sagas/update_actions.ts:
```diff
@@ -373,30 +400,137 @@ export function updateVolumeTracing(
+
+export function updateUserBoundingBoxInSkeletonTracingAction(
+  boundingBoxId: number,
+  updatedProps: Partial<UserBoundingBox>,
```
should it be `PartialBoundingBoxWithoutVisibility` like in the sibling function?

philippotto reviewed (changes requested):
cool stuff :) I left a couple of comments for the front-end.

knollengewaechs commented on frontend/javascripts/viewer/model/helpers/compaction/compact_save_queue.ts:
```diff
@@ -70,6 +70,33 @@ function removeSubsequentUpdateNodeActions(updateActionsBatches: Array<SaveQueue
+function removeSubsequentUpdateBBoxActions(updateActionsBatches: Array<SaveQueueEntry>) {
+  const obsoleteUpdateActions = [];
+
+  // Actions are obsolete, if they are for the same bounding box and for the same prop.
+  // The given action is always compared to the next next one, as usually an
+  // updateUserBoundingBoxInSkeletonTracingAction and updateUserBoundingBoxInVolumeTracingAction
+  // is sent at the same time.
```
yes exactly! :white_check_mark:

knollengewaechs commented on frontend/javascripts/viewer/model/helpers/compaction/compact_save_queue.ts:
```diff
@@ -70,6 +70,33 @@ function removeSubsequentUpdateNodeActions(updateActionsBatches: Array<SaveQueue
+  // The given action is always compared to the next next one, as usually an
+  // updateUserBoundingBoxInSkeletonTracingAction and updateUserBoundingBoxInVolumeTracingAction
+  // is sent at the same time.
+  for (let i = 0; i < updateActionsBatches.length - 2; i++) {
+    const actions1 = updateActionsBatches[i].actions;
+    const actions2 = updateActionsBatches[i + 2].actions;
```
(and each of these actions is sent in a new batch)

knollengewaechs commented on frontend/javascripts/viewer/model/helpers/compaction/compact_save_queue.ts:
```diff
@@ -70,6 +70,33 @@ function removeSubsequentUpdateNodeActions(updateActionsBatches: Array<SaveQueue
+      actions1[0].value.boundingBoxId === actions2[0].value.boundingBoxId &&
+      _.isEqual(actions1[0].value.updatedPropKeys, actions2[0].value.updatedPropKeys)
+    ) {
+      obsoleteUpdateActions.push(updateActionsBatches[i]);
```
Okay, so I tried to do the following and it seems to work:

``` for (let i = 0; i < updateActionsBatches.length; i++) {
    if (i >= updateActionsBatches.length - 2) {
      relevantActions.push(updateActionsBatches[i]);
      continue;
    }
    const actions1 = updateActionsBatches[i].actions;
    const actions2 = updateActionsBatches[i + 2].actions;

    if (
      actions1.length !== 1 ||
      actions2.length !== 1 ||
      (actions1[0].name !== "updateUserBoundingBoxInSkeletonTracing" &&
        actions1[0].name !== "updateUserBoundingBoxInVolumeTracing") ||
      actions1[0].name !== actions2[0].name ||
      actions1[0].value.boundingBoxId !== actions2[0].value.boundingBoxId ||
      !_.isEqual(actions1[0].value.updatedPropKeys, actions2[0].value.updatedPropKeys)
    ) {
      relevantActions.push(updateActionsBatches[i]);
    }
  }
  return relevantActions; 
  ```
  
  I think this makes the code harder to understand because of the negations. But maybe thats just me. I thought about putting everything into one big negation but felt like this was easy to miss. I also think it might be a bit more prone to errors, as the last update actions have to be added manually in the first lines of the loop. Personally I would have liked the version before better, and this code is also more similar to the other methods in this module (e.g. `removeSubsequentUpdateSegmentActions`). But I understand that this is a weak argument.

What do you think?

philippotto replied:
hmm, you are right, it's not getting easier to understand. frankly, I'm a bit hesitant about this function. also the "+ 2" stuff (which I didn't understand initially in my other comment) is a bit weird (especially because it bakes in assumptions over how the update actions exist). for example, it would break if there is no volume tracing (or more than 1 volume tracing).
I think it would be best to have a call about this :) maybe we can avoid the need for this function altogether (e.g., by avoiding that resizing a bbox emits so many actions in the first place).

knollengewaechs commented on frontend/javascripts/viewer/model/sagas/update_actions.ts:
```diff
@@ -373,30 +400,137 @@ export function updateVolumeTracing(
-export function updateUserBoundingBoxesInSkeletonTracing(
-  userBoundingBoxes: Array<UserBoundingBox>,
+
+export function addUserBoundingBoxInSkeletonTracingAction(
```
that makes a lot of sense!

knollengewaechs commented on frontend/javascripts/viewer/model/sagas/update_actions.ts:
```diff
@@ -16,6 +17,8 @@ export type NodeWithTreeId = {
   treeId: number;
 } & Node;
 
+type PartialBoundingBoxWithoutVisibility = Partial<Omit<UserBoundingBox, "isVisible">>;
```
of course!

knollengewaechs commented on frontend/javascripts/viewer/view/version_entry.tsx:
```diff
@@ -76,13 +82,53 @@ const descriptionFns: Record<
+    description: `Toggled the visibility of bounding box ${firstAction.value.boundingBoxId}.`,
+    icon: <EditOutlined />,
+  }),
+  updateUserBoundingBoxVisibilityInVolumeTracing: (
+    firstAction: UpdateUserBoundingBoxInVolumeTracingAction,
+  ): Description => ({
```
thanks!

knollengewaechs commented on frontend/javascripts/viewer/model/sagas/update_actions.ts:
```diff
@@ -373,30 +400,137 @@ export function updateVolumeTracing(
   actionTracingId: string,
 ) {
   return {
-    name: "updateUserBoundingBoxesInSkeletonTracing",
```
okay, thanks for the hint!

knollengewaechs commented:
I think I addressed all the issues that came up during review and I also implemented a list of booleans that get send to the server. @MichaelBuessemeyer the JSON now looks like this: `{
    name: actionName,
    value: {
      boundingBoxId,
      actionTracingId,
      updatedProps,
      hasUpdatedBoundingBox,
      hasUpdatedName,
      hasUpdatedColor,
    },` (current naming version, maybe up for change).

philippotto commented on frontend/javascripts/viewer/view/version_entry.tsx:
```diff
@@ -76,14 +86,66 @@ const descriptionFns: Record<
+  updateUserBoundingBoxesInVolumeTracing: (
+    firstAction: LEGACY_UpdateUserBoundingBoxesInVolumeTracingUpdateAction,
+  ): Description => ({
+    description: `Updated bounding boxes ${firstAction.value.boundingBoxes.map((bbox) => bbox.id).join()}.`,
```
```suggestion
    description: `Updated bounding boxes ${firstAction.value.boundingBoxes.map((bbox) => bbox.id).join(", ")}.`,
```

philippotto commented on frontend/javascripts/test/model/compact_save_queue.spec.ts:
```diff
@@ -0,0 +1,403 @@
+        timestamp: 1747994347590,
+        authorId: "123",
+        actions: [
+          {
+            name: "updateUserBoundingBoxInVolumeTracing",
+            value: {
+              boundingBoxId: 3,
+              actionTracingId: "volumeTracing1",
+              updatedProps: {
+                boundingBox: {
+                  topLeft: [3224, 3955, 944],
+                  width: 1,
+                  height: 157,
+                  depth: 12,
+                },
+              },
+              hasUpdatedBoundingBox: true,
+              hasUpdatedName: false,
+              hasUpdatedColor: false,
+            },
+          },
```
could you use the `updateUserBoundingBoxInVolumeTracing` action creator here (and in the rest of the file)? I think, the code will become more concise and we would also test the logic in the action creator a bit.

philippotto commented on frontend/javascripts/test/model/compact_save_queue.spec.ts:
```diff
@@ -0,0 +1,403 @@
+            branchPointCount: 0,
+          },
+        },
+        info: '["INITIALIZE_CONNECTOME_TRACING","SET_TD_CAMERA_WITHOUT_TIME_TRACKING * 5","SET_INPUT_CATCHER_RECTS","SET_TD_CAMERA_WITHOUT_TIME_TRACKING","UPDATE_CONNECTOME_FILE_LIST","SET_MAXIMUM_ZOOM_FOR_ALL_MAGS_FOR_LAYER * 4","SET_TOOL","CHANGE_USER_BOUNDING_BOX * 2","SET_TD_CAMERA_WITHOUT_TIME_TRACKING","CHANGE_USER_BOUNDING_BOX * 3"]',
```
the info stuff can be empty for the purpose of the test

fm3 reviewed (approved):
Backend LGTM

MichaelBuessemeyer commented:
before merging @philippotto did you test this? At least I did not :see_no_evil:
